### DISCUSSION TITLE
Is jsoncons thread-safe?

### DISCUSSION THREAD

**Discussion by recsater** [danielaparker/jsoncons#486] [Q&A] [answered]
I have not encountered any issues while using 20 cores, but I would like an official response. Thank you.

**Comment by danielaparker** [comment]
A `basic_json` has the same thread safety guarantees as the standard library container classes:

- there can be multiple concurrent readers of the same `basic_json` object
- if there is one writer, there can be no other writers and no readers

The `json_schema` extension has a `make_schema` function that produces a compiled representation of a JSON Schema, and the, `jsonpath` and `jmespath` extensions have  `make_expression` functions that produce a compiled representation of JSONPath and JMESPath expressions. These compiled representations can be shared by multiple threads.

**Reply by recsater** [reply]
Is there no issue when encoding or decoding the same struct simultaneously across multiple threads? not basic_json object

**Reply by danielaparker** [reply, marked as answer]
Should be fine, there's no shared mutable state.
